docs/proposals/templates.md

```diff
@@ -0,0 +1,595 @@
+  possible to write a Mustache template which, after Mustache processing, resulted in a Template which could then be instantiated 
+  through the normal template instantiating process.
+
+*Parameterization should not compromise reproducibility. For instance, it should be possible to manage template arguments as well as 
+templates under version control.*
+
+* Templates are a single file, including default or chosen values for parameters.  They can easily be managed under version control.
+
+*It should be possible to specify template arguments (i.e., parameter values) declaratively, in a way that is "self-describing" 
+(i.e., naming the parameters and the template to which they correspond). It should be possible to write generic commands to 
+process templates.*
+
+* Parameter definitions include metadata which describes the purpose of the parameter.  Since parameter definitions are part of the template,
+  there is no need to indicate which template they correspond to.
+
+*It should be possible to validate templates and template parameters, both values and the schema.*
+
+* Template objects are subject to standard api validation.
+
+*It should also be possible to validate and view the output of the substitution process.*
+
+* The `/processedTemplates` api returns the result of the substitution process, which is itself a Template object that can be validated.
+
+*It should be possible to generate forms for parameterized templates, as discussed in #4210 and #6487.*
+
+* Parameter definitions provide metadata that allows for the construction of form-based UIs to gather parameter values from users.
+
+*It shouldn't be inordinately difficult to evolve templates. Thus, strategies such as versioning and encapsulation should be 
+encouraged, at least by convention.*
+
+* Templates can be versioned via annotations on the template object.
+
+## Key discussion points
+The preceding document is opinionated about each of these topics, however they have been popular topics of discussion so they are called out explicitly below.
+
+### Where to define parameters
+
+There has been some discussion around where to define parameters that are being injected into a Template
+
+1. In a separate standalone file
+2. Within the Template itself
+
+This proposal suggests including the parameter definitions within the Template, which provides a self-contained structure that
+can be easily versioned, transported, and instantiated without risk of mismatching content.  In addition, a Template can easily
+be validated to confirm that all parameter references are resolveable.
+
+Separating the parameter definitions makes for a more complex process with respect to
+* Editing a template (if/when first class editing tools are created)
+* Storing/retrieving template objects with a central store
+
+Note that the `/processedTemplates/sometemplate` subresource would accept a standalone set of parameters to be applied to `sometemplate`.
```

Comment:
The subresource I imagined was something like:

/apis/extensions/v1beta1/namespaces/myns/templates/mytemplate/substituted